engines/agi/keyboard.cpp

```diff
@@ -97,54 +307,128 @@ int AgiEngine::doPollKeyboard() {
+		// special menu controller slot found
+		if (controllerSlotSpecial != controllerSlotESC) {
+			// not the same as the ESC slot (is the same in Manhunter AppleIIgs, we need to replace "pause"
+			if (controllerSlotSpecial >= 10) {
+				// slot needs to be at least 10
+				// Atari ST SQ1 maps the special key, but doesn't trigger any menu with it
+				// the controller slot in this case is 8.
+				return controllerSlotSpecial;
+			}
+		}
+	}
+	return -1;
+}
+
+bool AgiEngine::handleController(uint16 key) {
+	ScreenObjEntry *screenObjEgo = &_game.screenObjTable[SCREENOBJECTS_EGO_ENTRY];
+
+	if (key == 0) // nothing pressed
+		return false;
+
+	// This previously skipped processing, when ESC was pressed and called menu directly.
+	// This original approach was bad, because games check different flags before actually allowing the
+	//  user to enter the menu. We checked a few common flags, like for example the availability of the prompt.
+	//  But this stopped the user being able to enter the menu, when the original interpreter actually allowed it.
+	//  We now instead implement this feature using another way for those platforms.
+	if (key == AGI_KEY_ESCAPE) {
+		// Escape pressed, user probably wants to trigger the menu
+		// For PC, just passing ASCII code for ESC will normally trigger a controller
+		//  and the scripts will then trigger the menu
+		// For other platforms, ESC was handled by platforms to trigger "pause game" instead
+		// We need to change ESC to a platform specific code to make it work properly.
+		//
+		// There are exceptions though. Mixed Up Mother Goose on AppleIIgs for example actually sets up
+		//  ESC for pause only. That's why we also check, if the key is actually mapped to a controller.
+		// For this special case, we actually replace the pause function with a menu trigger.
+		// Replacing "pause" all the time wouldn't work out as well, becaue games like KQ1 on Apple IIgs
+		//  actually disable "pause" when ego has been killed, which means we wouldn't be able to access
+		//  the menu anymore in that case.
+		if (_menu->isAvailable()) {
+			// menu is actually available
+			if (_game.specialMenuTriggerKey) {
+				int16 specialMenuControllerSlot = getSpecialMenuControllerSlot();
+
+				if (specialMenuControllerSlot >= 0) {
+					// menu trigger found, trigger it now
+					_game.controllerOccured[specialMenuControllerSlot] = true;
+					return true;
+				}
+			}
+			// Otherwise go on and look for the ESC controller
+		}
+	}
+
 
 	// AGI 3.149 games, The Black Cauldron and King's Quest 4 need KEY_ESCAPE to use menus
 	// Games with the GF_ESCPAUSE flag need KEY_ESCAPE to pause the game
-	if (key == 0 ||
-		(key == KEY_ESCAPE && getVersion() != 0x3149 && getGameID() != GID_BC && getGameID() != GID_KQ4 && !(getFeatures() & GF_ESCPAUSE)) )
-		return false;
+	//		(key == KEY_ESCAPE && getVersion() != 0x3149 && getGameID() != GID_BC && getGameID() != GID_KQ4 && !(getFeatures() & GF_ESCPAUSE)) )
```

Comment:
yes, i kept it in for now. But the original approach wasn't that good anyway. I can't test Black Cauldron because I don't own it. Menu works with the new method in KQ4.